### ISSUE TITLE
[AArch64 & QEMU] GDB and Qemu crashes together while connecting to QUEMU gdb stub

### ISSUE THREAD

**Issue by TheAifam5** [pwndbg/pwndbg#570]
### Description
I seen similar issue like https://github.com/pwndbg/pwndbg/issues/85.
I tested with the official one `aarch64-linux-gnu-gdb` and the `gdb-multiarch` package and the same issue occurs.

After connecting the GDB to the QEMU, they crash with `segmentation fault` and `abort` errors.

### Steps to reproduce
- Start QEMU using: ` qemu-system-aarch64 -M raspi3 -kernel kernel8.img -d in_asm -S -s` (you need an image, for example from here: https://github.com/bztsrc/raspi3-tutorial)
- Start GDB using `aarch64-linux-gnu-gdb` and type `target remote localhost:1234`

### My setup
OS: ArchLinux x64
show version
```
GNU gdb (GDB) 8.2
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-pc-linux-gnu --target=aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
```

py import sys; print(sys.version):
```
3.7.1 (default, Oct 22 2018, 10:41:28) 
[GCC 8.2.1 20180831]
```

pwndbg version/git commit id: 19eaa19
```
pwndbg> version
Python Exception <class 'ValueError'> invalid literal for int() with base 10: 'warning:': 
Error occurred in Python command: invalid literal for int() with base 10: 'warning:'
```

QEMU log:
```
qemu-system-aarch64 -M raspi3 -kernel kernel8.img -d in_asm -S -s
qemu-system-aarch64: /build/qemu/src/qemu-3.0.0/hw/dma/bcm2835_dma.c:146: bcm2835_dma_read: Assertion `size == 4' failed.
[1]    10577 abort (core dumped)  qemu-system-aarch64 -M raspi3 -kernel kernel8.img -d in_asm -S -s
```

GDB Log:
```
pwndbg> target remote localhost:1234
Remote debugging using localhost:1234
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x0000000000000000 in ?? ()
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/elf.py", line 318, in iter_phdrs
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 81, in update
    page  = pwndbg.memory.Page(start, stop-start, 6 if not is_executable() else 7, 0, '[stack]')
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 131, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
RuntimeError: generator raised StopIteration

If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome
> /usr/share/pwndbg/pwndbg/stack.py(131)is_executable()
-> for phdr in pwndbg.elf.iter_phdrs(ehdr):
(Pdb) 
(Pdb) c
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/elf.py", line 318, in iter_phdrs
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 173, in caller
    raise e
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 81, in update
    page  = pwndbg.memory.Page(start, stop-start, 6 if not is_executable() else 7, 0, '[stack]')
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 131, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
RuntimeError: generator raised StopIteration
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/elf.py", line 318, in iter_phdrs
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 131, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
RuntimeError: generator raised StopIteration

If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome
> /usr/share/pwndbg/pwndbg/stack.py(131)is_executable()
-> for phdr in pwndbg.elf.iter_phdrs(ehdr):
(Pdb) c
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/elf.py", line 318, in iter_phdrs
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 173, in caller
    raise e
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/stack.py", line 131, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
RuntimeError: generator raised StopIteration
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 116, in explore_registers
    find(pwndbg.regs[regname])
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 75, in find
    return explore(address)
  File "/usr/share/pwndbg/pwndbg/abi.py", line 165, in caller
    return func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 105, in explore
    page = find_boundaries(address_maybe)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 366, in find_boundaries
    start = pwndbg.memory.find_lower_boundary(addr)
  File "/usr/share/pwndbg/pwndbg/memory.py", line 357, in find_lower_boundary
    pwndbg.memory.read(addr, 1)
  File "/usr/share/pwndbg/pwndbg/memory.py", line 44, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.error: Remote connection closed

If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome
> /usr/share/pwndbg/pwndbg/memory.py(44)read()
-> result = gdb.selected_inferior().read_memory(addr, count)
(Pdb) c
Traceback (most recent call last):
  File "/usr/share/pwndbg/pwndbg/events.py", line 173, in caller
    raise e
  File "/usr/share/pwndbg/pwndbg/events.py", line 169, in caller
    func()
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 116, in explore_registers
    find(pwndbg.regs[regname])
  File "/usr/share/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 75, in find
    return explore(address)
  File "/usr/share/pwndbg/pwndbg/abi.py", line 165, in caller
    return func(*args, **kwargs)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 105, in explore
    page = find_boundaries(address_maybe)
  File "/usr/share/pwndbg/pwndbg/vmmap.py", line 366, in find_boundaries
    start = pwndbg.memory.find_lower_boundary(addr)
  File "/usr/share/pwndbg/pwndbg/memory.py", line 357, in find_lower_boundary
    pwndbg.memory.read(addr, 1)
  File "/usr/share/pwndbg/pwndbg/memory.py", line 44, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.error: Remote connection closed
[1]    7286 segmentation fault (core dumped)  aarch64-linux-gnu-gdb
```

**Comment by bash-c** [comment]
I meet something like this. I got the following when debugging a mips binary
```bash
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
  0x400000   0x402000 r-xp     2000 0      /root/mipstake
  0x402000   0x411000 ---p     f000 1000   /root/mipstake
  0x411000   0x412000 rw-p     1000 1000   /root/mipstake
0x7f7ca000 0x7f7ee000 r-xp    24000 0      [linker]
0x7f7ca000 0x7f7ee000 r-xp    24000 0      /etc/qemu-binfmt/mips/lib/ld.so.1
0x7f7ee000 0x7f7fd000 ---p     f000 23000  [linker]
0x7f7ee000 0x7f7fd000 ---p     f000 23000  /etc/qemu-binfmt/mips/lib/ld.so.1
0x7f7fd000 0x7f7fe000 r--p     1000 23000  [linker]
0x7f7fd000 0x7f7fe000 r--p     1000 23000  /etc/qemu-binfmt/mips/lib/ld.so.1
0x7f7fe000 0x7f7ff000 rw-p     1000 24000  [linker]
0x7f7fe000 0x7f7ff000 rw-p     1000 24000  /etc/qemu-binfmt/mips/lib/ld.so.1
0x7fffe000 0x80000000 rwxp     2000 0      [stack]
pwndbg> b *0x400A78
Breakpoint 1 at 0x400a78
pwndbg> c
Continuing.
warning: GDB can't find the start of the function at 0x7f7ce403.

    GDB is unable to find the start of the function at 0x7f7ce403
and thus can't determine the size of that function's stack frame.
This means that GDB may be unable to access that stack frame, or
the frames below it.
    This problem is most likely caused by an invalid program counter or
stack pointer.
    However, if you think GDB should simply search farther back
from 0x7f7ce403 for code which looks like the beginning of a
function, you can increase the range of the search using the `set
heuristic-fence-post' command.
/build/gdb-GT4MLW/gdb-8.1/gdb/thread.c:93: internal-error: thread_info* inferior_thread(): Assertion `tp' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.

This is a bug, please report it.  For instructions, see:
<http://www.gnu.org/software/gdb/bugs/>.

Aborted (core dumped)
```

At first I thinks there is something wrong with gdb-multiarch(`GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git`), but finally I found that using gdb-multiarch without any plugin woks well.

**Comment by CarloMara** [comment]
Same problem here,

@TheAifam5  by any chance does your elf have any custom sections or non standard relocations? Mine does and I'm suspecting that's the problem, as works with _normal_ elfs.

Carlo

**Comment by CarloMara** [comment]
So, this is probably a common case of every day bit rot.

[PEP479](https://www.python.org/dev/peps/pep-0479/) changed the way `StopIteration` works, as now it raises a `RuntimeError` when triggered. This of course is non backwards compatible, but it was fairly trivial to to fix, as the PEP author took grate care to illustrate how to deal with old broken code.

I'll send a PR to get this fixed, I'm really wondering tho how this wasn't caught earlier, maybe I'm grossly missing something.

Carlo

**Comment by disconnect3d** [comment]
This issue should be fixed with #613.

**Comment by TheAifam5** [comment]
@CarloMara its not a “elf” format. Its just a kernel from scratch for Raspi3. I provided a link to the tutorial which has a LD file and prebuild image if you interested.
